### PR TITLE
Unblock relaxed write consistency level

### PR DESCRIPTION
**Pull request by takayuki**
This pull request contains two commits; one introducing an option in cmd/influx to select write consistency level, which has been set to 'any' by default, and another fixing writes possibly blocked even with relaxed write consistency level (e.g. any, one, or quorum).

Assume you have a 3-node cluster and one of those nodes went down. Write requests with write consistency level more relaxed than, or equal to quorum, should be completed immediately and successfully. But, the requests are possibly blocked waiting the dead node to respond.

### CODE REVIEW

**Comment by beckettsean** [comment]
@corylanou

File: cluster/points_writer.go [hunk 1]
```diff
@@ -274,32 +278,32 @@ func (w *PointsWriter) writeToShard(shard *meta.ShardInfo, database, retentionPo
 	var wrote int
 	timeout := time.After(w.WriteTimeout)
 	var writeError error
-	for _, nodeID := range shard.OwnerIDs {
+	for i := 0;  i < len(shard.OwnerIDs); i++ {
```
Comment (corylanou) [inline cluster/points_writer.go]: Why did this change?  I don't see us using `i`, but I could be missing it.  The original is much clearer and idiomatic in Go.

If it was to stop using the `nodeID`, you can simply do:
```go
for range shard.OwnerIDS {
```

**Comment by corylanou** [comment]
Overall, looks good.  I would like to see a test or two added.  One for testing the new flag added and that it parses correctly.  Would like one more set of eyes as well to the change on the cluster points writer.  Perhaps @otoolep or @jwilder

**Comment by otoolep** [comment]
Thanks @takayuki 

@jwilder definitely needs to see this. I understand the change to the write path code, but @jwilder needs to confirm that it is the right thing to do.

File: cluster/points_writer.go (hunk 1 above)
Comment (jwilder) [inline cluster/points_writer.go]: Agreed.  The `AsyncWriteResult` and for loop change is unnecessary.

**Comment by jwilder** [comment]
Moving the `write >= required` to the upper loop is the correct fix.  I was able to verify that a slow writer could cause a write of `any`, `one` or `quorum` to block and return a timeout even after it successfully wrote to the required nodes.  If the node was down (an we could not connect), it would return as quickly as possible in the current code.

**Comment by jwilder** [comment]
@takayuki If you could rebase, revert the looping change, and make sure you've signed the CLA.  This change should be good to go.  Thanks for the fix!

**Comment by takayuki** [comment]
Thank you @jwilder Would you spend some more with me, letting me think about the concern I originally had?

There are two loops involved here, one for starting goroutines and another for waiting for them to complete their tasks. The problem is, I thought, the second loop might be getting the results in a different order from one in which goroutines started. That's why I had to add AsyncWriteResult handling out-of-order arrivals.

Here is a conceptual example, taking the suggestion from @corylanou; the first loop starts node#1, then node#2, whereas the second loop is going to get the results from node#2, then node#1. I thought PointsWriter would have to take it into account.

```
	ownerIDs := [2]int{1,2}
	ch := make(chan int, len(ownerIDs))
	for _, nodeID := range ownerIDs {
		go func(nodeID int) {
			// node#1 sleeps for 10 secs, node#2 for 5 secs,
			// so node#2 is going to finish earlier.
			time.Sleep(time.Duration(10/nodeID)*time.Second)
			ch <- nodeID
		}(nodeID)
	}
	for range ownerIDs {
		select {
		case nodeID := <-ch:
			// node#2 arrives earlier, then node#1 later.
			fmt.Println(nodeID)
		}
	}
```

**Comment by otoolep** [comment]
Yeah, that's what I thought too @takayuki -- the new loop made sense to me.

**Comment by otoolep** [comment]
I've confirmed @takayuki has signed the CLA.

**Comment by jwilder** [comment]
@takayuki @otoolep The order in which the response arrive does not matter from a write perspective.  This code is writing to all shard owners in parallel and returning as soon as it receives the required number of successful responses.    

It does look like there is a bug in the logging here though: https://github.com/influxdb/influxdb/pull/3901/files#diff-99b560e650ccc1fa3f20c23df5a9b41bL287 which might be what the looping change was intended to fix.  It could log the wrong `NodeID` if the results arrive out of order.  That is not right so I think your looping change makes sense in light of that issue.

**Comment by jwilder** [comment]
Thanks @takayuki !

**Comment by otoolep** [comment]
Yes, your second point is the one @takayuki and I were making, it was just a logging thing.

**Pull request merged**